### DISCUSSION TITLE
How to make macOS RustDesk client starts at boot (on login screen)?

### DISCUSSION THREAD

**Discussion by C-Duv** [rustdesk/rustdesk#7762] [Q&A] [answered]
I am trying to deploy RustDesk remotely, using SSH (on multiple computers) and the goal is to allow remote access event when no one is physically present in front of the computer.

I managed to install it (using the `.dmg`, `/usr/bin/hdiutil attach` and copying to `/Applications`) and define macOS's *privacy settings*.

I created the `/Library/LaunchDaemons/com.carriez.RustDesk_service.plist` and `/Library/LaunchAgents/com.carriez.RustDesk_server.plist` files following [`src/platform/privileges_scripts/install.scpt`](https://github.com/rustdesk/rustdesk/blob/edb5529df3b505941d8fcb27d818b21b652bcaaf/src/platform/privileges_scripts/install.scpt).

But, after a reboot (to check if everything starts smoothly and unattended) I:

* do have `/Applications/RustDesk.app/Contents/MacOS/RustDesk --server` running as `root` user (checked using SSH, without any user physically logged in)
* cant get the session ID using `/Applications/RustDesk.app/Contents/MacOS/RustDesk --get-id`
* cannot connect to RustDesk session from another computer.

If I run `launchctl load -w /Library/LaunchDaemons/com.carriez.RustDesk_service.plist` I have more process (according to `ps aux | grep -i MacOS/RustDesk`) but same issue (not able to connect):

```
root  217  […]  2:33PM  0:03.09 /Applications/RustDesk.app/Contents/MacOS/RustDesk --server
root  537  […]  2:40PM  0:00.01 /bin/sh -c sleep 1.5; /Applications/RustDesk.app/Contents/MacOS/RustDesk --service
root  539  […]  2:40PM  0:01.01 /Applications/RustDesk.app/Contents/MacOS/RustDesk --service
```

I tried with both versions v1.2.3-2 and nightly (`65f7541ec76`): same issue.

Any idea?

macOS version is 13.5 ("Ventura")

**Comment by rustdesk** [comment]
The main problem is macOS does not allow to grant “Screen Recording" permission from command line.
https://rustdesk.com/docs/en/client/mac/

> #Make MacOS trust RustDesk.app
> xattr -rd com.carriez.RustDesk /Applications/RustDesk.app

**Reply by C-Duv** [reply]
I think I got that part covered, using [*jacobsalmela/tccutil*](https://github.com/jacobsalmela/tccutil) I managed to allow `kTCCServiceScreenCapture` (and others) to RustDesk.

Anyway, would it prevent ~macOS~ RustDesk to start at boot?

**Reply by rustdesk** [reply]
Try out `--install-service` (it does [src/platform/privileges_scripts/install.scpt](https://github.com/rustdesk/rustdesk/blob/edb5529df3b505941d8fcb27d818b21b652bcaaf/src/platform/privileges_scripts/install.scpt)), but you'd better use nightly build.

https://github.com/rustdesk/rustdesk/blob/edb5529df3b505941d8fcb27d818b21b652bcaaf/src/platform/macos.rs#L141


https://github.com/rustdesk/rustdesk/blob/ae23f0de03903dcc2c541856a842ed79404f21a9/src/core_main.rs#L242

**Reply by C-Duv** [reply]
I tried `/Applications/RustDesk.app/Contents/MacOS/RustDesk --install-service` with both 1.2.3-2 and 1.2.4 and service is not started immediately nor after a reboot.

`/var/root/Library/Logs/RustDesk/install-service/RustDesk_rCURRENT.log` shows:

```
[2024-04-18 18:10:10.790016 +02:00] INFO [src/core_main.rs:164] main start args:["--install-service"]
[2024-04-18 18:10:10.790396 +02:00] INFO [src/core_main.rs:243] start --install-service
```

**Reply by C-Duv** [reply]
When both `RustDesk --server` and `RustDesk --service` are run by `root` connection does not work.

Once an user (say `admin`) physically logs in, `root`'s `RustDesk --server` is killed and another `RustDesk --server` is started under the user's account, but connection does not work either.

I tried to manually install nightly v1.2.4 (`65f7541ec76`) on macOS v14.4.1 ("Sonoma") as a basic end user and it does not work either:

1. Download the [`rustdesk-1.2.4-aarch64.dmg`](https://github.com/rustdesk/rustdesk/releases/download/nightly/rustdesk-1.2.4-aarch64.dmg)
2. Open the .dmg and install it's `RustDesk.app` to `/Applications`
3. Start RustDesk
4. Go to _Settings_ > _General_ and make sure the service is running 
5. Make sure remote RustDesk connection to the computer (from another computer) using displayed session ID and password works
6. Close RustDesk
7. Reboot the computer

Remote RustDesk connection to the computer using displayed session ID and password from step 5 should work.

So, my question is does RustDesk even support "automatic startup" on macOS?

This is the last missing feature for our deployment .Self-hosted pro server configuration works fine, so does strategies, address book and groups. It's just weird it can't run before login.


During my initial tests I think I had managed to have RustDesk (v1.2.3 or v1.2.3-1) running at logon screen: authenticating (as a macOS user) would close the RustDesk session but client would eventually re-connect in 2-3 seconds and I could see macOS's user desktop. Or was it with AnyDesk (we tried both)?

**Reply by rustdesk** [reply]
BTW, --service run as Launch Daemon with root, --server run as LauchAgent with normal user.

![image](https://github.com/rustdesk/rustdesk/assets/71636191/2d700b43-fdfb-4902-9e4e-d4215e9710e1)

**Reply by C-Duv** [reply]
I got that, but:

* do I need to run CLI command(s) to tell RustDesk to start on boot? Or should it work right away after installation?
* if so, which command(s)?

As I want it to work without any user logged-in I guess any command I need to type would only be as `root`, and not via any "normal user".

**Reply by rustdesk** [reply]
no need to run CLI. `--install-service` does the same thing, but I have never run it with `root`. Could you share your deployment scripts with https://github.com/jacobsalmela/tccutil etc? I will take a look and test once I have time after I finish msi/android custom client.

**Reply by C-Duv** [reply]
I'll compile my tccutil commands and post them here.

If I read you well, a "normal" install (manually, under user _foo_ for example) should enable itself to start on machine boot and allow remote connection without any user logged-in physically?

**Reply by C-Duv** [reply]
Here are my `tccutil` commands.

They insert a rule for RustDesk Client and then enable them (=allow) in various privacy services.

```Shell
bundleId='com.carriez.rustdesk'

sudo tccutil --service kTCCServiceSystemPolicyAllFiles --insert "${bundleId}"
sudo tccutil --service kTCCServiceSystemPolicyAllFiles --enable "${bundleId}"

sudo tccutil --service kTCCServiceScreenCapture --insert "${bundleId}"
sudo tccutil --service kTCCServiceScreenCapture --enable "${bundleId}"

sudo tccutil --service kTCCServiceListenEvent --insert "${bundleId}"
sudo tccutil --service kTCCServiceListenEvent --enable "${bundleId}"

sudo tccutil --service kTCCServiceAccessibility --insert "${bundleId}"
sudo tccutil --service kTCCServiceAccessibility --enable "${bundleId}"
```

Note: Requires `tccutil` at commit `512b611` (or later) which includes two required fixes:

* Fix for the « Attempting to write a readonly database.  You probably need to disable SIP. » error message.
* Use `auth_value=2` when enabling service (needed by macOS 12+) ([issue #61](https://github.com/jacobsalmela/tccutil/issues/61)).

**Reply by rustdesk** [reply]
Thanks, i will test boot later. Or we might learn from TV, https://www.teamviewer.com/en-us/global/support/knowledge-base/teamviewer-classic/deployment/mass-deployment-on-macos/

**Reply by C-Duv** [reply]
By looking for a workaround, I noticed RustDesk can accept incoming connection on the "locked screen" if it was initially started by an user which locked it's computer (does not matter if someone connected to the computer between RustDesk startup and macOS lock).

So, on computer _C1_:
* Start computer
* Log-in physically using user _U1_
* Launch RustDesk from GUI
* Lock the computer session (Apple menu > Lock Screen)

Then, on computer _C2_: you can RustDesk-connect to computer _C1_, the screen will display macOS's lock screen and you'll be able to type the _U1_'s password and access it's macOS desktop/session.

But you can't log-in with another user account (say _U2_) than _U1_ (the one that initially started RustDesk), well you can but you'll see _U1_ desktop and won't be able to interact with it.
The physical screen of _C1_ will display _U2_'s macOS desktop/session.

**Reply by C-Duv** [reply]
I looked at AnyDesk, it works flawlessly on log-in screen and the only running process (at boot) is `/Applications/AnyDesk.app/Contents/MacOS/AnyDesk --service` (under `root` user).

I did the same, executed `/Applications/RustDesk.app/Contents/MacOS/RustDesk --service` as `root` and I got the following output and no return to shell (process continuously running):

> 2024-04-23 18:53:06.664 RustDesk[419:3760] XType: Using static font registry.

But: no sign-in on the RustDesk Server, no connection possible, and `lsof -i -P | grep LISTEN` shows no RustDesk opened port

**Comment by C-Duv** [comment]
So… @rustdesk How can we, via CLI using administrator account, enable RustDesk to start at boot (**boot**, not user **login**) after having installed it (id. copied the `RustDesk.app` folder to `/Applications`)?

The use case is deployment of the application on multiple machines for remote-hand assistance of non-tech users, it should be fairly simple:

1. Connect as administrator via SSH
2. Install the application
3. Configure it (set some options in `/Library/Preferences/com.carriez.RustDesk/*.toml`)
4. Enable it for boot startup
5. Start it now (avoids waiting for next startup)

Somehow the RustDesk integration on macOs makes the last 2/3 points difficult…

**Reply by C-Duv** [reply]
Sorry about the @ I wasn't sure you'll get notified somehow (and it just happen I've updated 2 issues the same day).

But you were the one marking it answered (over my `tccutil` commands) in the first place.

And, excuse me but: how could you not know about this aspect of the application? You are the main (sole ?) maintainer of RustDesk.

The https://rustdesk.com/docs/en/self-host/client-deployment/ page gives some command for deploying:

```
# Run the rustdesk command with --get-id and store the output in the rustdesk_id variable
cd /Applications/RustDesk.app/Contents/MacOS/
rustdesk_id=$(./RustDesk --get-id)

# Apply new password to RustDesk
./RustDesk --server &
/Applications/RustDesk.app/Contents/MacOS/RustDesk --password $rustdesk_pw &> /dev/null

/Applications/RustDesk.app/Contents/MacOS/RustDesk --config $rustdesk_cfg

# Kill all processes named RustDesk
rdpid=$(pgrep RustDesk)
kill $rdpid &> /dev/null
```

But ends with a step involving GUI:

```
echo "Please complete install on GUI, launching RustDesk now."
open -n /Applications/RustDesk.app
```

The application seems to be able to do what I've described above if clicking on the GUI (I've ended up having it working), so it's obviously also possible using CLI.
I am not asking you to do it, just asking for pointers on technical aspects I have to look into in order to obtain my desired behavior.

Such as:
* Dealing with 2 `.plist` configuration file: a *LaunchDaemon* and a *LaunchAgent*
  Which one starts what, which one is the "listening to incoming connections" binary and how to start/stop it.
* Dealing with 2 `.toml` configuration file: `RustDesk.toml` and `RustDesk2.toml`
* Dealing with configuration files being duplicated for every system user

Unless those internal architecture details of implementation are documented somewhere I can't find a better person to ask to if not you.

**Reply by rustdesk** [reply]
All code here,

https://github.com/rustdesk/rustdesk/blob/251e1a3487518f763e0bb2b7eafb505fbd75b56d/src/platform/macos.rs#L193

simply say it only has two lines
1, run the script with root
2, start the launch agent with normal user

**Comment by PC-Support-Services** [comment]
Having RD as a pkg **_installation_** file would / should solve these problems.

**Comment by Kerrnel** [comment]
A script to setup daemons - tested on a Proxmox MacOS VM (15.2) [ Update - fixed USE_PW ]

[rustdesk-daemon-install.gz](https://github.com/user-attachments/files/18205585/rustdesk-daemon-install.gz)

**Reply by rustdesk** [reply]
``` bash
#!/bin/bash
#
#	rustdesk-daemon-install - public domain 2024

#	Ver	Who	Date	Note
#	001	GK	241219	Initial setup - tested on ProxMox machine - permissions had to be manually responded to in GUI


#	Setup RustDesk services for Mac - even in a VM environment (presuming a copy of Preferences from another machine)

#	Optionally set ID and password - to make it work on a MacOS Vm

#	Server is for a user
#	Service is for the machine

#
# This script will find RustDesk.app in /Applications or a subfolder or subfolder of a subfolder
#

# From lib/log
Msg()	{	echo "$(date +%d:%k%M%S)	$$	$@" >&2;	}
Err()	{	Msg "<***> $@ <***>"; return 42;			}
Out()	{	local l; while read l; do Msg "$l"; done;	}
Throw()	{	[ -n "$1" ] && Msg "=== $1 ==="; exit ${2:-22}; exit 9;	}

# Are we running as root - not really normal for a mac - maybe sudo'd
[ $(id -u) -eq 0 ] && \
	{ SUDO=''; FORUSER=${SUDO_USER:-$USER}; } || \
	{ SUDO='sudo'; FORUSER=$(id -un); }

FORUID=$(id -u "$FORUSER")
DEFAULTPATH='/Applications/RustDesk.app'
APP=''

AGENT=''	# Assets from Github
DAEMON=''
ASCPT=''

USESCRIPT=0		#	Use AppleScript instead of direct install
USE_ID=''		#	Set ID of this machine via id= in toml file
USE_PWD=''		#	Set permanent password to this if non-null

CARRIER='com.carriez.RustDesk'
agent_path="/Library/LaunchAgents/${CARRIER}_server.plist"
daemon_path="/Library/LaunchDaemons/${CARRIER}_service.plist"
root_pref_path="/var/root/Library/Preferences/$CARRIER"
user_pref_path=~/"Library/Preferences/$CARRIER"

pref_files='RustDesk.toml RustDesk2.toml'

GetAssets()
 {
	local i
	Msg "Verify $FORUSER ($FORUID) preferences..."
	for i in $pref_files; do
		[ ! -s "$user_pref_path/$i" ] && Throw "Missing preference file $i for user $FORUSER - this is needed for the system-wide settings copy

	If running this in a VM
		Copy a known good $CARRIER
			from ~/Library/Preferences
			on a real mac
			(or a Linux .config/$CARRIER)
			to $user_pref_path on this VM 

		Then run this script again an use the -i ID option to
			set the ID of this RustDesk node.
"
	done

	Msg 'Loading assets from github'
	AGENT=$(curl -s 'https://raw.githubusercontent.com/rustdesk/rustdesk/master/src/platform/privileges_scripts/agent.plist' | sed -e "s|$DEFAULTPATH|$APP|g")
	DAEMON=$(curl -s 'https://raw.githubusercontent.com/rustdesk/rustdesk/master/src/platform/privileges_scripts/daemon.plist' | sed -e "s|$DEFAULTPATH|$APP|g")
	ASCPT=$(curl -s 'https://raw.githubusercontent.com/rustdesk/rustdesk/master/src/platform/privileges_scripts/install.scpt')
	[ -z "$AGENT" -o -z "$DAEMON" -o -z "$ASCPT" ] && Throw "Error getting one or more assets - url probably changed. (A/D/S sizes: ${#AGENT}/${#DAEMON}/${#ASCPT})"
 }

InstallViaAppleScript()
 {
	Msg "Current User:	$FORUSER"

	# This script is pretty stoopid. It does not get the home directory e.g. by algorithm
	$SUDO osascript -e "$ASCPT" "$DAEMON" "$AGENT" "$FORUSER"
 }

KickstartServerGUI()
 {
	if [ -s "$agent_path" ]; then
		Msg "Kickstart server for $FORUID ($FORUSER)) - if it fails, user probably not logged in"
		$SUDO launchctl enable "gui/$FORUID/${CARRIER}_server"
		$SUDO launchctl kickstart -kp "gui/$FORUID/${CARRIER}_server"
	else
		Err "Ooops - where is $agent_path"
	fi
 }

SetPassword()
 {
	if [ -n "$USE_PWD" ]; then
		# 2024-12-19 - Happy Birthday Nik - currently RustDesk checks for the current executable being in /Applications/*.app/... - therefore
		#	If $APP is not in /Applications directly, we have to symlink to it temporarily
		if [ "${APP#/Applications/}" != 'RustDesk.app' ]; then
			Msg "Temporary symlink for password"
			$SUDO ln -s "$APP" '/Applications/RustDesk.app'
		fi
		Msg "Setting Password"
		'/Applications/RustDesk.app/Contents/MacOS/RustDesk' --password "$USE_PWD"
		if [ "${APP#/Applications/}" != 'RustDesk.app' ]; then
			Msg "Removing temporary symlink for password - you may want to create one anyway - other things may assume /Applications as the location of RustDesk.app"
			$SUDO rm '/Applications/RustDesk.app'
		fi
	fi
 }

InstallViaShell()
 {
	Msg "Install agent:	$agent_path"
	echo "$AGENT" | $SUDO tee "$agent_path" >/dev/null

	Msg "Install daemon:	$daemon_path"
	echo "$DAEMON" | $SUDO tee "$daemon_path" >/dev/null

	# Should already be - but incase they were created by user error
	$SUDO chown 0:0 "$agent_path" "$daemon_path"

	local i
	$SUDO mkdir -p "$root_pref_path"
	for i in $pref_files; do
		if [ "$i" = 'RustDesk.toml' -a "$USE_ID" != '' ]; then
			# Set id and allow the service (as of current version 2024-12-19) to reencrypt and stash in enc_id
			while [ ${#USE_ID} -lt 6 ]; do
				Err "$USE_ID is too short (I think) - prefixing hostname"
				USE_ID="$(uname -n)-$USE_ID"
			done
			sed -I '.bak' -e "s/^id = .*/id = '$USE_ID'/"  -e "s/^enc_id = .*/id = '$USE_ID'/" "$user_pref_path/$i"
		fi
		$SUDO cp -a "$user_pref_path/$i" "$root_pref_path"
	done

	SetPassword

	Msg "Unload / Reload Daemon"
	local d="${daemon_path##*/}" # No path
	d="${d%.*}"	# No plist
	$SUDO launchctl stop "$d" >/dev/null 2>&1
	$SUDO launchctl remove "$d" > /dev/null 2>&1
	$SUDO launchctl load -w "$daemon_path"

	Msg "Unload / Reload Agent for $FORUSER"
	local a="${agent_path##*/}" # No path
	a="${a%.*}"	# No plist
	launchctl stop "$a" >/dev/null 2>&1
	launchctl remove "$a" > /dev/null 2>&1
	launchctl load "$agent_path"
 }

FindApp()
 {
	# Find RustDesk.app up to two levels deep in /Applicatins
	local d
	APP=''
	for d in /Applications /Applications/* /Applications/*/*; do if [ -d "$d" ]; then
		[ -d "$d/RustDesk.app" ] && { APP="$d/RustDesk.app"; Msg "Found app:	$APP"; break; }
	fi; done
 }

IsAgentLoaded()
 {
	local p
	p=$(launchctl list | while read pid sta com; do
			[ "$com" = "$1" ] && { echo "$pid:$sta"; break; }
		done)
	[ -n "$p" ]
 }

Help()
 {
	Throw "

Usage: ${0##*/} [options]

	-a	Show Agent plist contents
	-d	Show Daemon plist contents
	-h	This help
	-i id	Set RustDesk ID of this machine
	-p pwd	Set RustDesk permanent password
	-s	Show Script contents
	-S	Use AppleScript to install - not recommended
	-v	List the Agent/Daemon plists installed
"
 }


# Begin Script Body
Msg "##################################################################"
Msg "# $(date) | Starting ${0##*/}"
Msg "#*								*#"
Msg	"#**      If prompted for a password - that is proably sudo     **#"
Msg "##################################################################"

FindApp
Msg "##################################################################"
GetAssets

while getopts 'adehi:p:sStv' o; do case "$o" in
	'a')	echo "$AGENT"; Throw 'Done.'	;;
	'd')	echo "$DAEMON"; Throw 'Done.'	;;
	'h')	Help							;;
	'i')	USE_ID="$OPTARG"				;;
	'p')	USE_PWD="$OPTARG"				;;
	's')	echo "$ASCPT"; Throw 'Done.'	;;
	'S')	USESCRIPT=1						;;
	'v')	ls -l "$agent_path" "$daemon_path"; 	Throw 'Done.'	;;
esac; done; shift $(($OPTIND - 1))

while [ ! -d "$APP" ]; do
	Msg 'Could not find app installed in /Applications/... - sleeping 30'
    sleep 30
    FindApp
done

IsAgentLoaded 'com.carriez.RustDesk_service' && \
	Throw  'com.carriez.RustDesk_service is already loaded'

Msg "******* STARTING SERVICE INSTALLATION *******"
[ $USESCRIPT -eq 1 ] && InstallViaAppleScript || InstallViaShell
KickstartServerGUI

Msg 'Done. Note - if you move the app, you will have to reinstall or edit the LaunchDaemon and LaunchAgent plists'
ls -l "$agent_path" "$daemon_path"
```

**Reply by fufesou** [reply]
Here's an improved version compared to the one above.  https://github.com/rustdesk/rustdesk/discussions/7762#discussioncomment-11612098

https://github.com/rustdesk/rustdesk/wiki/macOS-Auto%E2%80%90Start-Service-Setup-(for-Remote---MDM-Deployment)

**Comment by jelockwood** [comment]
I have recently been looking at how best to automate installing RustDesk for Mac and have found the following issues and workarounds.

Firstly I have to say that whilst RustDesk as a remote control app seems a good solution its approach for configuring on a Mac is awful and totally anti-Mac. It in this area appears to be a terrible reskinned Linux app. (This is purely in relation to configuring it.) As a result it does not use standard macOS plist files and hence cannot be configured via standard Mac tools or an MDM. Worse still it appears RustDesk requires its proprietary settings be installed in a users individual Preferences folder and not e.g. /Library/Preferences

I have therefore - so far done the following.

1. Use an MDM to push PPPC settings to enable Accessibility and enable allowing a _non-admin_ level user to turn on Screen Recording and Input Recording (as sadly required by Apple a local logged in user still has to turn this on themselves but with the PPPC settings they don't have to use an admin login to authorise it)
2. Use a standard Mac enterprise tool such as [Munki](https://github.com/munki/munki) or an MDM or other, to deploy the RustDesk app to each client Mac, I use Munki which copies the dmg to the client Mac, mounts it and copies the app to the /Applications folder
3. Use a standard Mac enterprise tool such as Munki or an MDM or other to deploy the [Outset](https://github.com/macadmins/outset) utility to each client Mac, I use Munki which copies the pkg to the client Mac and then runs it

_The above works for both non-admin and admin level users_

4. As an admin on an administrator Mac _manually_ import the setting string into a copy of RustDesk, this creates the (ugh!) proprietary format preference files in ~/Library/Preferences/com.carriez.rustdesk
5. I then created a base64 encoded tar of the ~/Library/Preferences/com.carriez.rustdesk folder which has all the required settings to allow a RustDesk Mac client to talk to the open source server on premise
6. I then included the base64 encoded tar in a shell script using 'doc here' aka 'cat eof' format as per the approach I detail [here](https://jelockwood.blogspot.com/2018/06/using-cat-eof-in-shell-script-to.html)
7. This script containing the encoded RustDesk preferences then needs to be copied to /usr/local/outset/login-once/ on each client Mac, the Outset documentation details how to generate an installer pkg to do this which Munki (or an MDM) could then deploy to each client Mac, the Outset utility will then run this script _once_ for each new user who logins _on that computer_ the script will then restore the preferences to ~/Library/Preferences/com.carriez.rustdesk and as this script is executed _as the user_ it is created and owned _by that user_ in the users home directory.

_The above also works for both non-admin and admin level users_

To summarise - the RustDesk app is deployed in a way that does _not_ require an admin user account, the Outset utility is deployed in a way that does _not_ require an admin user account, the RustDesk preferences are deployed in a way that works for any user and the user does _not_ require an admin level account.

Note: The RustDesk supplied Mac deployment script has two bugs. Firstly it _has to_ be run by an admin level user and _requires_ manual entry of the users password, secondly if the ~/Library/Preferences/com.carrierz.rustdesk folder does not already exist, it is created in the right place but it is set with access to _only_ root and hence the user - even if an admin level user cannot access the _generated_ preferences. This permissions/ownership issue does not occur if rather than using their script you manually import/paste the config string.

My goal was to automate deploying the app, automate deploying the settings and to make this all work for non-admin level user logins.

Due to Apple restrictions it is not possible to auto-enable screen recording. I am aware of the custom tccutil solutions but do not want to install them and do not want to turn off SIP protection.

Since the 'official' Mac deployment script has the above issues and hence is literally useless - although for admin level users a simple single line addition to the script to 'fix' the preference folder permissions would help for them, the above is an approach that so far seems to act as a successful replacement and works also for non-admin users. Via Munki (and [AutoPkg](https://github.com/autopkg/autopkg)) it has the bonus of also automatically updating the RustDesk desktop app.

Munki could be set to automatically run either a post-install script to create launchctl files or run another pkg to do this. As all our user Macs are FileVault encrypted I had no need to ensure RustDesk was available at the login window since this would not work till after a FileVault enabled user has logged in anyway.

**Comment by Mahdi-Shabani** [comment]
Thanks for the details. On macOS, just having “--server” running as root is not enough. The ID is owned/managed by the background service, and the desktop-capture server must run in a GUI session (including the LoginWindow session). If the service can’t register to the ID server, “--get-id” will return nothing and you can’t connect.

What usually goes wrong

The service is not loaded as a LaunchDaemon (or exits), so no ID is registered.
The server is started in the wrong session (system/root) instead of a GUI session (Aqua/LoginWindow).
Server addresses are not configured (ID/Relay), or firewall blocks them.
First‑run permissions (Screen Recording/Accessibility/Incoming connections) were never granted.
How to fix/check

A) Make sure the service is running (system domain)

It must be a LaunchDaemon:
sudo launchctl bootstrap system /Library/LaunchDaemons/com.carriez.RustDesk_service.plist
sudo launchctl enable system/com.carriez.RustDesk_service
sudo launchctl kickstart -k system/com.carriez.RustDesk_service
The plist should run:
ProgramArguments = [ "/Applications/RustDesk.app/Contents/MacOS/service" ]
with KeepAlive/RunAtLoad set, and (optional) StandardOutPath/StandardErrorPath for logs.
Check status/logs:
sudo launchctl print system/com.carriez.RustDesk_service
sudo log show --last 10m --process RustDesk

B) Start the desktop server in a GUI session

LaunchAgents don’t run until a GUI session exists. To support the login screen, the LaunchAgent should include:
LimitLoadToSessionType = LoginWindow
ProgramArguments = [ "/Applications/RustDesk.app/Contents/MacOS/RustDesk", "--server" ]
KeepAlive/RunAtLoad = true
Then load it into the LoginWindow domain:
sudo launchctl bootstrap loginwindow /Library/LaunchAgents/com.carriez.RustDesk_server.plist
You can also load it for a specific user session (after login):
sudo launchctl bootstrap gui/$(id -u someuser) /Library/LaunchAgents/com.carriez.RustDesk_server.plist
C) Configure server addresses (ID/Relay)
If you use your own servers, preconfigure them so the service can register and get an ID without any user logged in. Create RustDesk2.toml (system‑wide) for the service, for example:
/Library/Application Support/rustdesk/RustDesk2.toml
[server]
id = "YOUR_ID_SERVER_HOST_OR_IP"
relay = "YOUR_RELAY_HOST_OR_IP:PORT"

api/key if you use them
If you rely on the public servers, you can skip this, but make sure outbound connections are allowed.

D) Network/firewall

Ensure outbound traffic to your ID/relay is allowed (often TCP 21115/21116/21117, or put relay on 443/TCP to pass corporate/5G).
On the Mac itself, if Application Firewall is on, make sure RustDesk is allowed for incoming connections (the first run prompt appears only in a GUI session).
E) Generate/verify ID
Once the service is up and registered, this should return the ID (even with no user logged in):
/Applications/RustDesk.app/Contents/MacOS/RustDesk --get-id
If it’s empty, check the service logs for connection errors to the ID server.

Quick checklist

Service: active in system domain, logs show connected to hbbs (ID server).
Server: loaded into LoginWindow (or GUI user) domain.
TCC: Screen Recording + Accessibility approved for RustDesk.
Network: relay reachable (test from the Mac: nc -vz RELAY_HOST 443).
If you want, I can look at your two plist files. Post them (redact any sensitive info) plus the output of:

sudo launchctl print system/com.carriez.RustDesk_service
sudo log show --last 5m --process RustDesk
/Applications/RustDesk.app/Contents/MacOS/RustDesk --get-id
Once the service successfully registers, you’ll get a stable ID and should be able to connect at the login screen.

**Comment by rkaalma** [comment]
A bit late to the party but..

My main goal was to get some sort of remote desktop solution working for a 2023 Mac Mini (M2 cpu). I tried countless approaches, tried to make it work with Gemini deep-research, etc - always the same result - device only accessible after manual login. Got tired, ditched the RustDesk and went for TeamViewer - initially was the same (no access after reboot) until I found someone pointing out that **"when the FileVault disc encryption is enabled, then programs can only start after logging in"**. I turned it off for my device and device was available after reboot - problem solved for me. I'm sure that turning off the FileVault disc encryption isn't the ideal solution but for me the Mac is more of a "testing device".

I didn't have the time to test it out with RustDesk after cleaning it up but hopefully this information is useful in some way - if not, feel free to delete my comment (it that's a thing in GH).

**Reply by MyCodeArea** [reply]
That works! Thanks! 
I am running here to leave these words for the later people who seeing this.

I am using windows to connect my MAC mini through RustDesk. When FileVault turned off, after rebooting of MAC OS, I can connect to MAC and enter the password, login into MAC in RustDesk window. That's really convenient. I won't have to connect a keyboard and monitor to do this. I can complete work on windows, and don something when I need to do on MAC. 

But, there is an issue: when we login MAC in RustDesk, the connection will disconnect. And then we will have to reconnect. I think this might be the restart of RustDesk on MAC when login. This only happens after reboot of MAC. If MAC recover from sleep, this will not happen. 
That is not perfect, but not bother much.

**Reply by xlionjuan** [reply]
> But, there is an issue: when we login MAC in RustDesk, the connection will disconnect. And then we will have to reconnect. I think this might be the restart of RustDesk on MAC when login. This only happens after reboot of MAC. If MAC recover from sleep, this will not happen.
That is not perfect, but not bother much.

Same as Linux, it is expected behaviour

**Reply by C-Duv** [reply]
Could the behavior changed to allow/offer a seamless login process?

**Comment by rustdesk** [comment, marked as answer]
A full wiki is here https://github.com/rustdesk/rustdesk/wiki/macOS-Auto%E2%80%90Start-Service-Setup-(for-Remote---MDM-Deployment)

**Reply by ambsalem1** [reply]
This works great.  Thanks!

**Reply by ambsalem1** [reply]
Here's what I did for PPPC (Privacy Preferences Policy Control) in Intune:

1. Created new MacOS Intune Policy for Settings Catalog
2. Add Settings: Privacy > Privacy Preferences Policy Control > Screen Capture
     Allow (Deprecated): REMOVE THIS OPTION!
     Authorization: Allow
     Code Requirement: HZF9JMC8YN
                  This is standard for all RustDesk clients. To find, open terminal: codesign -display -r - /Applications/RustDesk.appz
     Identifier: bundle ID
     Identifier Type: com.company.clientname
     Static Code: true

3. Assign policy to devices
